### PR TITLE
[V1] Add circuits

### PR DESCRIPTION
**Pull request by 0xSachinK**
- The new proof is for the following statement: `I on-ramper, have an email from Venmo, signed by their DNS key, that states that someone sent $X amount of USD off-chain to a Venmo ID XYZ, and wishes to release the escrowed funds for order ID ABC on-chain.`
- Nullifier: Email hash
- To prevent frontrunning: OrderID is added as an input to the circuit.
- Update regex circuits
    - Amount (Extracted from email subject)
    - Off-ramper user ID (Extracted from body and hashed)
    - Remove old regexes.
- Adds `order_id` as an input to the circuit
- Fix generate input error when calling from the UI. The reason is because the `DkimVerifier` function is making a call to Google DNS provider (`resolveDNSHTTP`) in the UI vs command line. In the `resolveDNSHTTP` function in `dkim/tools.js` we are extracting the `0` index from the Answer array which is not the right public key. The right public key is in the `1` index of the array. This is because Venmo DNS has 2 elements in the array while Twitter (ZK-Email default example) only has 1
  - Venmo DNS https://dns.google/resolve?name=224i4yxa5dv7c2xz3womw6peuasteono._domainkey.amazonses.com&type=TXT
  - Twitter DNS https://dns.google/resolve?name=dkim-201406._domainkey.twitter.com&type=TXT

### Circuit stats
```
[INFO]  snarkJS: Curve: bn-128
[INFO]  snarkJS: # of Wires: 6400562
[INFO]  snarkJS: # of Constraints: 6618823
[INFO]  snarkJS: # of Private Inputs: 7478
[INFO]  snarkJS: # of Public Inputs: 18
[INFO]  snarkJS: # of Labels: 30385320
[INFO]  snarkJS: # of Outputs: 4
```

### Stats
- Compilation time: 563.87s
- Proving key gen time on server: 782s
- Witness gen: 60s
- Proof generation: 9.2s

### TODO for future PR
- Add more regex security. Update regex circuit to include the carriage values to prevent content spoofing attacks "\r\n"
- Trusted setup scripts

### CODE REVIEW

File: circuit/VenmoMMV1_regex.circom [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+		eq[3][i] = IsEqual();
+		eq[3][i].in[0] <== in[i];
+		eq[3][i].in[1] <== 100;
+		and[3][i] = AND();
+		and[3][i].a <== states[i][3];
+		and[3][i].b <== eq[3][i].out;
+		states[i+1][4] <== and[3][i].out;
+		eq[4][i] = IsEqual();
+		eq[4][i].in[0] <== in[i];
+		eq[4][i].in[1] <== 101;
```
Comment (richardliang) [inline circuit/VenmoMMV1_regex.circom]: Regex parsing seems wrong. [47, 99,111,100,101] is `/code` so its not starting at the beginning of the preselector string

**Review by richardliang — CHANGES_REQUESTED** [review]

File: circuit/circuit_js/witness_calculator.js [hunk 2]
```diff
@@ -149,7 +149,7 @@ class WitnessCalculator {
 		throw new Error(`Too many values for input signal ${k}\n`);
 	    }
             for (let i=0; i<fArr.length; i++) {
-                const arrFr = toArray32(normalize(fArr[i],this.prime),this.n32)
+                const arrFr = toArray32(BigInt(fArr[i])%this.prime,this.n32)
```
Comment (0xSachinK) [inline circuit/circuit_js/witness_calculator.js]: Why edit this?

Comment (richardliang) [reply circuit/circuit_js/witness_calculator.js]: didn't make any edits here, auto generated?

File: circuit/scripts/compile_circuit.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
 cd circuit
 # Compile the circuit
-circom circuit.circom --r1cs --wasm --sym
+circom circuit.circom --wasm --r1cs --sym
```
Comment (richardliang) [inline circuit/scripts/compile_circuit.sh:3]: Note: reminder to run **without** --O1 flag both locally and in server

**Pull request merged**